apps/web/src/components/tx/confirmation-views/useTxPreview.ts

```diff
@@ -21,10 +32,40 @@ const useTxPreview = (
   const address = customSafeAddress ?? safeAddress
   const { operation = Operation.CALL, data = '', to, value } = safeTxData ?? {}
 
-  return useAsync(() => {
+  return useAsync(async () => {
     if (skip) return
-    return getTxPreview(chainId, address, operation, data, to, value)
-  }, [skip, chainId, address, operation, data, to, value])
+    const txPreview = await getTxPreview(chainId, address, operation, data, to, value)
+
+    // Validate that the tx preview matches the txData
+    // First of all the top level data needs to match:
+    const matchesSafeTxData =
+      sameAddress(txPreview.txData.to.value, safeTxData.to) &&
+      Number(txPreview.txData.operation) === Number(safeTxData.operation) &&
+      txPreview.txData.value === safeTxData.value &&
+      (txPreview.txData.hexData ?? '0x') === safeTxData.data
+
+    if (!matchesSafeTxData) {
+      throw new InvalidPreviewError("SafeTx data does not match the preview result's transaction data")
+    }
+
+    // validate the decodedData
+    const dataDecoded = txPreview.txData.dataDecoded
+
+    if (dataDecoded) {
+      const abiString = `function ${dataDecoded.method}(${dataDecoded.parameters?.map((param) => param.type).join(',')})`
+      const abiInterface = new Interface([abiString])
+      const rawDataFromDecodedData = abiInterface.encodeFunctionData(
+        dataDecoded.method,
+        dataDecoded.parameters?.map((param) => param.value),
+      )
+
+      if (rawDataFromDecodedData !== safeTxData.data) {
+        throw new InvalidPreviewError('Decoded data does not match raw data')
+      }
+    }
+
+    return txPreview
+  }, [skip, chainId, address, operation, data, to, value, safeTxData])
```

Comment:
I suggest we extract this into a separate hook. Apart from `useTxPreview`, which is used only when _creating_ a tx, we should call this hook also when confirming proposed txs – in that case they use TransactionDetails.